### PR TITLE
WIP: adding inline style flag

### PR DESCRIPTION
**Pull request by iruzevic**
# Description

- added global window object data set for everything.
- added a bunch of helper to get data from the window object.
- added inline style tag feature

# Screenshots / Videos

https://user-images.githubusercontent.com/23283324/152498069-50cada56-45ef-4458-a2be-8b1231429796.mov

# Note
This must not be merged until PHP part is done

### CODE REVIEW

File: schemas/globalManifest.json [hunk 1]
```diff
@@ -25,6 +25,16 @@
 			"pattern": "^#(?:[0-9a-fA-F]{3}){1,2}$",
 			"description": "The same as background color, foreground represents the namespace secondary color. You can override this in every block manifest.json."
 		},
+		"config": {
+			"type": "object",
+			"description": "With config object you can flag features can be active or not depending on your need.",
+			"properties": {
+				"outputVariablesInline": {
+					"type": "string",
+					"description": "This key will output all inline style tags used for css variables in one style tag at the top of the head element. This is used for optimisation purposes. Default: true"
```
Comment (volfkarlo) [inline schemas/globalManifest.json:34]:
```suggestion
					"description": "This key will output all inline style tags used for CSS variables in one style tag at the top of the head element. This is used for optimization purposes. Default: true"
```

File: scripts/editor/get-manifest-details.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * @param {string} componentName Component name to search.
+ *
+ * @returns {object}
+ */
+export const getComponentManifest = (componentName) => {
+	return getComponentsManifest().filter((item) => item.componentName === componentName) ?? {};
+};
+
+/**
+ * Returns css variables array.
```
Comment (volfkarlo) [inline scripts/editor/get-manifest-details.js:87]:
```suggestion
 * Returns CSS variables array.
```

File: scripts/editor/index.js [hunk 3]
```diff
@@ -8,6 +8,21 @@ export {
...
+	setConfigOutputCssVariables,
+	getGlobalManifest,
+	getBlocksManifest,
+	getBlockManifest,
+	getComponentsManifest,
+	getComponentManifest,
+	getStyles,
+	getWrapperManifest,
+	getNamespace,
+	getBlockFullName
```
Comment (volfkarlo) [inline scripts/editor/index.js:24]:
```suggestion
	getBlockFullName,
```

File: scripts/editor/output-css-variables.js [hunk 4]
```diff
@@ -421,6 +428,12 @@ export const outputCssVariables = (attributes, manifest, unique, globalManifest,
 					${variable.join('\n')}
 					}
 				`;
+
+				styles.variables.push({
+					type: '',
+					variable,
+					value
```
Comment (volfkarlo) [inline scripts/editor/output-css-variables.js:435]:
```suggestion
					value,
```

File: scripts/index.js [hunk 5]
```diff
@@ -39,6 +39,21 @@ export {
...
+	setConfigOutputCssVariables,
+	getGlobalManifest,
+	getBlocksManifest,
+	getBlockManifest,
+	getComponentsManifest,
+	getComponentManifest,
+	getStyles,
+	getWrapperManifest,
+	getNamespace,
+	getBlockFullName
```
Comment (volfkarlo) [inline scripts/index.js:55]:
```suggestion
	getBlockFullName,
```

**Review by volfkarlo — CHANGES_REQUESTED** [review]
Few trailing commas, and typos. Other than that 🎉

**Review by volfkarlo — DISMISSED** [review]

**Review by iobrado — APPROVED** [review]

File: blocks/init/src/Blocks/manifest.json [hunk 6]
```diff
@@ -1,8 +1,11 @@
 {
-	"$schema": "https://raw.githubusercontent.com/infinum/eightshift-frontend-libs/develop/schemas/settings.json",
+	"$schema": "https://raw.githubusercontent.com/infinum/eightshift-frontend-libs/develop/schemas/globalManifest.json",
 	"namespace": "eightshift-boilerplate",
 	"background": "#FBF9FF",
 	"foreground": "#9973E3",
+	"config": {
+		"outputVariablesInline": true
```
Comment (goranalkovic-infinum) [inline blocks/init/src/Blocks/manifest.json:7]: Maybe we should change the key name to `outputCssVariablesGlobally` or reverse the logic of the current one, that makes a tad more sense (inline would mean that they’re _inline_ with the components) 😄

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
Left one suggestion, nice job!

File: blocks/init/src/Blocks/manifest.json (hunk 6 above)
Comment (dingo-d) [reply blocks/init/src/Blocks/manifest.json:7]: I agree with this. First part that I've noticed was the missing `Css` in the name, and the second one was the inline part that confused me (that's what we're doing by default).

File: schemas/globalManifest.json [hunk 7]
```diff
@@ -25,6 +25,16 @@
 			"pattern": "^#(?:[0-9a-fA-F]{3}){1,2}$",
 			"description": "The same as background color, foreground represents the namespace secondary color. You can override this in every block manifest.json."
 		},
+		"config": {
+			"type": "object",
+			"description": "With config object you can flag features can be active or not depending on your need.",
```
Comment (dingo-d) [inline schemas/globalManifest.json:30]:
```suggestion
			"description": "With config object, you can flag features that can be active or not depending on your need.",
```

File: scripts/editor/output-css-variables.js [hunk 8]
```diff
@@ -456,8 +496,68 @@ export const outputCssVariables = (attributes, manifest, unique, globalManifest,
...
+					outputItem += `@media (${type}-width: ${value}px) {
+							${variable.join('\n')}
+						}
+					`;
+				}
+			}
+
+			output += `.${name}[data-id='${unique}'] {
+				${outputItem}
+				}
```
Comment (dingo-d) [inline scripts/editor/output-css-variables.js:543]:
```suggestion
			output += `.${name}[data-id='${unique}'] {
					${outputItem}
				}
```

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left some minor comments, looks good otherwise 🙂

File: blocks/init/src/Blocks/manifest.json (hunk 6 above)
Comment (iruzevic) [reply blocks/init/src/Blocks/manifest.json:7]: agreed i will put `outputCssVariablesGlobally` instead

Comment (iruzevic) [reply blocks/init/src/Blocks/manifest.json]: updated

File: scripts/editor/get-manifest-details.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * Returns one block manifest details.
+ *
+ * @param {string} blockName Block name to search.
+ *
+ * @returns {object}
+ */
+export const getBlockManifest = (blockName) => {
+	return getBlocksManifest().filter((item) => item.blockName === blockName) ?? {};
```
Comment (volfkarlo) [inline scripts/editor/get-manifest-details.js:63]: Does this return array or an object? 

I would say that 
`.find` would be a better fit.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

```suggestion
	return getBlocksManifest().find((item) => item.blockName === blockName) ?? {};
```

File: scripts/editor/get-manifest-details.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * Returns one component manifest details.
+ *
+ * @param {string} componentName Component name to search.
+ *
+ * @returns {object}
+ */
+export const getComponentManifest = (componentName) => {
+	return getComponentsManifest().filter((item) => item.componentName === componentName) ?? {};
```
Comment (volfkarlo) [inline scripts/editor/get-manifest-details.js:83]: The same as above. The return type doesn't seem appropriate.

**Review by volfkarlo — CHANGES_REQUESTED** [review]
Micro changes. Sorry for not spotting it earlier 😓

File: scripts/editor/register-blocks.js [hunk 11]
```diff
@@ -613,7 +614,18 @@ export const registerBlocks = (
...
 
-	window['eightshift'][process.env.VERSION] = componentsManifest;
+	window['eightshift'][process.env.VERSION] = {
+		globalManifest,
+		components: componentsManifest,
+		blocks: blocksManifests,
+		wrapper: wrapperManifest,
+		styles: [],
+		config: {
+			outputVariablesInline: true,
```
Comment (gabriel-glo) [inline scripts/editor/register-blocks.js:624]: Also here `outputCssVariablesGlobally`

**Review by gabriel-glo — CHANGES_REQUESTED** [review]
Few questions. Looks good

**Review by mbmjertan — DISMISSED** [review]
Apart from others have already commented, looks good to me. Nice work!

File: scripts/editor/get-manifest-details.js (hunk 9 above)
Comment (iruzevic) [reply scripts/editor/get-manifest-details.js:63]: i will check if this changes anything

**Review by simun-ivanac — APPROVED** [review]

**Review by gabriel-glo — APPROVED** [review]

**Review by iobrado — APPROVED** [review]

**Pull request closed without merging**